### PR TITLE
Slight changes

### PR DESCRIPTION
**Pull request by jmmelko**
* Added support for parsing note repetition coded as 'xN' => a xN text is appended to the chord image in the HTML output => this functionnality is not advertised yet in the prompt
* Changed the shift (transposition) from full octave to individual note shift. So the user must now enter +7 or -7 to shift an entire octave. + corrected a bug where note_shift was undefined
* Changed the way of splitting notes in chords with Western notation. Instead of assuming that notes are always 2 char-long, I split before the note uppercase name. => adds support for sharps and flats + note repetition (xN)

PS: it seems that I don't have write access to the beta branch of the main project.
It is fine with me, but this requires Tracey to acknowledge my pull requests.

PPS: Alexdolly gave me permission to use SVG + a package to convert to JPG. So maybe I will first try to output a SVG and using pyCairo for conversion.

### CODE REVIEW

**Comment by t1-tracey** [comment]
Ohh these sound great. Sorry I was unfamiliar with GitHub settings, I have just found the option to give you write access. Are you able to merge now?

**Comment by t1-tracey** [comment]
@jmmelko I downloaded the beta branch today since I had some time, and was curious to test the note shift :smile: 

hmm oh I see transposition doesn't seem to be implemented yet, that's ok. I had some suggestions and I would be happy to try help with the implementation of these, just need to try understand first~

1. about the note_shift-

> Changed the shift (transposition) from full octave to individual note shift. So the user must now enter +7 or -7 to shift an entire octave. + corrected a bug where note_shift was undefined

- I have a suggestion to use +12 or -12 for shifting by semitones. Would it be easier to implement semitones than accounting for a mix of whole steps and half steps? (in +7 to -7). And I sometimes play in the in-between scales like Ab major for let it go, and E major for a Journey song (parts of which I've shared sheets before) 

`['C4', 'Db4', 'D4', 'Eb4', 'E4', 'F4', 'Gb4', 'G4', 'Ab4', 'A4', 'Bb4', 'B4', 'C5']` (labelled according to [scientific pitch notation](https://en.wikipedia.org/wiki/Scientific_pitch_notation) which I didn't realise I'd been making mistakes with on the wiki)

2. And wow I noticed it seems the program will try to make a list of the possible keys that the song could be in, from the notes. I think that's really cool, just not sure if something else might be more ux-friendly. I am impressed though.

- I wanted to suggest — do you think it might be easier to explicitly ask the user if there's a key they'd like to type in first? (if not, check for possible keys and ignore accidentals. otherwise default to C major) and then ask for note shift?

in my spare time, I'm happy to do the implementation of these^^

**Comment by jmmelko** [comment]
Helli Tracey,

Yes, feel free to do it, since I’m 100% working on the SVG part.

Well clearly the program was not intentionally thought for transposition, which is logical since in Sky’s world there is no semitones. It’s like trying to translate a rich language into a poor language; so we’re not far from your grammar ideas...
I didn’t do it because I didn’t find any satisfactory solution : Dictionary are cools but sometimes a bit overkill since transposing is just a circular shift. You may have noticed that I have already created long dictionaries to do the so-called “note shift”, which is actually a transposition in Sky’s limited language.

Sooo... maybe could you write a generic transposition function, working with any set of symbols:
* either the full tonal scale (12 semitones)
* or the reduced set of tones of Sky

 taking into account that:

* chromatic transposition is just a circular shift ;
* of course the modulo is 12 for the chromatic scale, I got mixed up with 7 because this is the modulo for Sky’s scales
* people can use b or #, I know there are different practices, but I don’t know them
* you may have to add or substract 12 to stay within the same pitch range (?)

If you do it, then most likely my “find_key“ procedure will be obsolete, but that’s ok.

Thankfully you checked the notation or else I would have guessed it was A4b.

Good luck ^^

JMM.

**Comment by t1-tracey** [comment]
Hi JMM,

Yeah you’re right that this program wasn’t really originally intended for transposition. I had noticed the dictionaries for note_shift, I thought it was cool how you used the tuples/coordinates^^.

Hm, I think — How about we decide transposition is outside the scope of this program? (Translating from one text to notation to another text notation can be the primary function of another program if someone would like to do it — and the result can be imported as a text file into this program. And this program would be reserved mostly for translating from text to visual.)

The scope for this program can be (for Western notation):

- the user can specify a recommended key (e.g. say the key is E major)
- we accept notation in that key, but ignore all accidentals outside that key. ([E4, F4 G#4/Ab4, A4, B4, C#5, D#5, E5, F#5, G#5, A5, B5, C#6, D#6, E6] would be accepted, but C4 would not be)
- the program will produce visual sheet music for playing in that key

Actually I like this idea so much, I think I’d like to pick out your code for exporting into text file/translating between text notations, and refactor it into a separate module. Thus the core of this program will be text-to-visual.

I’ll take some time to work on this

On 31 Oct 2019, at 3:53 am, jmmelko <notifications@github.com<mailto:notifications@github.com>> wrote:

Helli Tracey,

Yes, feel free to do it, since I’m 100% working on the SVG part.

Well clearly the program was not intentionally thought for transposition, which is logical since in Sky’s world there is no semitones. It’s like trying to translate a rich language into a poor language; so we’re not far from your grammar ideas...
I didn’t do it because I didn’t find any satisfactory solution : Dictionary are cools but sometimes a bit overkill since transposing is just a circular shift. You may have noticed that I have already created long dictionaries to do the so-called “note shift”, which is actually a transposition in Sky’s limited language.

Sooo... maybe could you write a generic transposition function, working with any set of symbols:
* either the full tonal scale (12 semitones)
* or the reduced set of tones of Sky

taking into account that:

* chromatic transposition is just a circular shift ;
* of course the modulo is 12 for the chromatic scale, I got mixed up with 7 because this is the modulo for Sky’s scales
* people can use b or #, I know there are different practices, but I don’t know them
* you may have to add or substract 12 to stay within the same pitch range (?)

If you do it, then most likely my “find_key“ procedure will be obsolete, but that’s ok.

Thankfully you checked the notation or else I would have guessed it was A4b.

Good luck ^^

JMM.

**Comment by jmmelko** [comment]
Tracey,

I think we agree; I’ll just rephrase it to be sure I understand everything (sorry...):
* we decide that transposition within the tonal scale (with semitones) is out of the scope of this program. Or at least out of the beta branch.
* however we can still keep transposition within Sky’s world (aka note shift), but let’s not call it transposition in the program to avoid confusing the user.

* the program already does what you propose, but for the C key only: it ignores accidentals outside the reduced C scale in Sky’s world (=outside the position_map dictionary), and then it transpose into a visual sheet in C4
* the idea would be to do the same for other keys available in Sky. This was what you intended first when asking to fill the “recommended key” field in the HTML file.

To do this:
* the program first detect playable keys (using my new function), to help the user decide
* then it asks in which key the input should be interpreted and the visual sheet rendered (so we just have to change the “No key found; assuming it is C” message in main.py to let the user decide whatever the output of the automatic key finder)
* then it does the usual Western->Sky translation but using a specific dictionary (position_map) for the chosen key
* of course if the input is already in Sky then we simply ask for the recommended Key.

In the end, the biggest step seems to decide between writing a dictionary for each possible key by hand, computationally writing the dictionaries (ie using python to generate python code, that’s possible!), or using another system: arrays, or a mix between arrays and dictionaries.

Have fun! :-)

jmmelko

**Comment by t1-tracey** [comment]
Hi jmmelko, I think it's fine to clarify^^, I will continue the discussion on #9
